From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: John Brooks <john.brooks@jollamobile.com>
Date: Thu, 31 Oct 2013 05:31:10 -0600
Subject: [PATCH] Manually time out HTTP requests after 30 seconds

In some cases, QNAM appears to not give a finished or error signal at any time.
To avoid having authentication requests that can get stuck, time out HTTP requests
manually after 30 seconds.
---
 src/base-plugin.cpp | 18 ++++++++++++++++++
 src/base-plugin.h   |  1 +
 2 files changed, 19 insertions(+)

diff --git a/src/base-plugin.cpp b/src/base-plugin.cpp
index aa2d639e78f14a87e266a4d58d272a203ba1ed6a..60a7958e207d619421e54c5ccef2c226c700df4d 100644
--- a/src/base-plugin.cpp
+++ b/src/base-plugin.cpp
@@ -32,6 +32,7 @@
 #include <QNetworkProxy>
 #include <QDateTime>
 #include <QCryptographicHash>
+#include <QTimer>
 
 using namespace SignOn;
 using namespace OAuth2PluginNS;
@@ -50,6 +51,7 @@ public:
 
     QNetworkAccessManager *m_networkAccessManager;
     QNetworkReply *m_reply;
+    QTimer m_timeout;
     mutable BasePlugin *q_ptr;
 }; //Private
 
@@ -60,6 +62,8 @@ BasePluginPrivate::BasePluginPrivate(BasePlugin *q):
     m_reply(0),
     q_ptr(q)
 {
+    m_timeout.setInterval(1000 * 30);
+    m_timeout.setSingleShot(true);
 }
 
 BasePluginPrivate::~BasePluginPrivate()
@@ -71,6 +75,7 @@ void BasePluginPrivate::disposeReply()
 {
     Q_Q(BasePlugin);
 
+    m_timeout.stop();
     if (m_reply != 0) {
         QObject::disconnect(m_reply, 0, q, 0);
         m_reply->deleteLater();
@@ -82,6 +87,7 @@ BasePlugin::BasePlugin(QObject *parent):
     QObject(parent),
     d_ptr(new BasePluginPrivate(this))
 {
+    QObject::connect(&d_ptr->m_timeout, SIGNAL(timeout()), SLOT(onPostTimeout()));
 }
 
 BasePlugin::~BasePlugin()
@@ -130,6 +136,8 @@ void BasePlugin::postRequest(const QNetworkRequest &request,
             this, SLOT(onNetworkError(QNetworkReply::NetworkError)));
     connect(d->m_reply, SIGNAL(sslErrors(QList<QSslError>)),
             this, SLOT(handleSslErrors(QList<QSslError>)));
+
+    d->m_timeout.start();
 }
 
 void BasePlugin::serverReply(QNetworkReply *reply)
@@ -157,6 +165,16 @@ void BasePlugin::onPostFinished()
     serverReply(reply);
 }
 
+void BasePlugin::onPostTimeout()
+{
+    Q_D(BasePlugin);
+
+    if (d->m_reply) {
+        TRACE() << "network request manually timed out";
+        d->m_reply->abort();
+    }
+}
+
 void BasePlugin::onNetworkError(QNetworkReply::NetworkError err)
 {
     Q_D(BasePlugin);
diff --git a/src/base-plugin.h b/src/base-plugin.h
index 9a727184f14e18d0aa1f51ff7d09561bb43763fe..d5de1a81b6dfd4be1b1da2cf56bd2663964c64e9 100644
--- a/src/base-plugin.h
+++ b/src/base-plugin.h
@@ -74,6 +74,7 @@ protected:
 
 protected Q_SLOTS:
     void onPostFinished();
+    void onPostTimeout();
     void onNetworkError(QNetworkReply::NetworkError err);
     virtual void handleSslErrors(QList<QSslError> errorList);
 
